include/cpp-sort/detail/merge_insertion_sort.h

```diff
@@ -0,0 +1,453 @@
+    // Relational operators
+
+    template<typename Iterator1, typename Iterator2>
+    auto operator<(const group_iterator<Iterator1>& lhs,
+                   const group_iterator<Iterator2>& rhs)
+        -> bool
+    {
+        return lhs.base() < rhs.base();
+    }
+
+    template<typename Iterator1, typename Iterator2>
+    auto operator<=(const group_iterator<Iterator1>& lhs,
+                    const group_iterator<Iterator2>& rhs)
+        -> bool
+    {
+        return lhs.base() <= rhs.base();
+    }
+
+    template<typename Iterator1, typename Iterator2>
+    auto operator>(const group_iterator<Iterator1>& lhs,
+                   const group_iterator<Iterator2>& rhs)
+        -> bool
+    {
+        return lhs.base() > rhs.base();
+    }
+
+    template<typename Iterator1, typename Iterator2>
+    auto operator>=(const group_iterator<Iterator1>& lhs,
+                    const group_iterator<Iterator2>& rhs)
+        -> bool
+    {
+        return lhs.base >= rhs.base();
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Arithmetic operators
+
+    template<typename Iterator>
+    auto operator+(group_iterator<Iterator> it, std::size_t size)
+        -> group_iterator<Iterator>
+    {
+        return it += size;
+    }
+
+    template<typename Iterator>
+    auto operator+(std::size_t size, group_iterator<Iterator> it)
+        -> group_iterator<Iterator>
+    {
+        return it += size;
+    }
+
+    template<typename Iterator>
+    auto operator-(group_iterator<Iterator> it, std::size_t size)
+        -> group_iterator<Iterator>
+    {
+        return it -= size;
+    }
+
+    template<typename Iterator>
+    auto operator-(const group_iterator<Iterator>& lhs, const group_iterator<Iterator>& rhs)
+        -> typename group_iterator<Iterator>::difference_type
+    {
+        return (lhs.base() - rhs.base()) / lhs.size();
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Construction function
+
+    template<typename Iterator>
+    auto make_group_iterator(Iterator it, std::size_t size)
+        -> group_iterator<Iterator>
+    {
+        return { it, size };
+    }
+
+    template<typename Iterator>
+    auto make_group_iterator(group_iterator<Iterator> it, std::size_t size)
+        -> group_iterator<Iterator>
+    {
+        return { it.base(), size * it.size() };
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Merge-insertion sort
+
+    template<
+        typename RandomAccessIterator,
+        typename Compare,
+        typename Projection
+    >
+    auto merge_insertion_sort_impl(RandomAccessIterator first, RandomAccessIterator last,
+                                   Compare compare, Projection projection)
+    {
+        // Cache all the differences between a Jacobsthal number and its
+        // predecessor that fit in 64 bits, starting with the difference
```

Comment:
Let's call J(n) the nth Jacobsthal number. Instead of storing J(0), J(1), J(2)... what I store is actually J(1) - J(0), J(2) - J(1), J(3) - J(2)... except that I don't actually need the first ones so everything is shifted and the first value stored is J(4) - J(3).